### ISSUE TITLE
JIT for arm

### ISSUE THREAD

**Issue by benmkw** [bytecodealliance/wasmtime#2735]
#### Feature

JIT for arm64

#### Benefit

x86 works and it would be nice to be able to use it on M1 mac as well

#### Implementation

it currently panics here: https://github.com/bytecodealliance/wasmtime/blob/5fecdfa49150e3304c1b949aab73bd4a0a02dbac/cranelift/jit/src/backend.rs#L183-L195 I'm unsure what needs to change/ whats blocking it

#### Alternatives

only make JIT available on x86

**Comment by cfallin** [comment]
@bnjbvr any thoughts on this? I'm not sure how PLTs ordinarily work on macOS/aarch64 but perhaps it's a simple addition?

**Comment by bjorn3** [comment]
As workaround you could disable the `is_pic` flag when creating the `TargetIsa`.

**Comment by bjorn3** [comment]
This is one of the plt entries of an executable I compiled on my phone:

```asm
90 01 00 B0    adrp x16, #0x31000
11 F2 46 F9    ldr  x17, [x16, #0xde0]
10 82 37 91    add  x16, x16, #0xde0
20 02 1F D6    br   x17
```

**Comment by benmkw** [comment]
> As workaround you could disable the is_pic flag when creating the TargetIsa.

Tried with https://github.com/bytecodealliance/cranelift-jit-demo
```diff
diff --git a/Cargo.toml b/Cargo.toml
index 756983f..f6dd93f 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -8,7 +8,8 @@ description = "Toy language implemented using cranelift-jit"
 edition = "2018"
 
 [dependencies]
-cranelift = "0.69.0"
-cranelift-module = "0.69.0"
-cranelift-jit = "0.69.0"
+cranelift = "0.72.0"
+cranelift-module = "0.72.0"
+cranelift-jit = "0.72.0"
+cranelift-native = "0.72.0"
 peg = "0.6"
diff --git a/src/jit.rs b/src/jit.rs
index 160ac36..ffcff04 100644
--- a/src/jit.rs
+++ b/src/jit.rs
@@ -26,7 +26,17 @@ pub struct JIT {
 
 impl Default for JIT {
     fn default() -> Self {
-        let builder = JITBuilder::new(cranelift_module::default_libcall_names());
+        let mut flag_builder = settings::builder();
+        // On at least AArch64, "colocated" calls use shorter-range relocations,
+        // which might not reach all definitions; we can't handle that here, so
+        // we require long-range relocation types.
+        flag_builder.set("use_colocated_libcalls", "false").unwrap();
+        flag_builder.set("is_pic", "false").unwrap();
+        let isa_builder = cranelift_native::builder().unwrap_or_else(|msg| {
+            panic!("host machine is not supported: {}", msg);
+        });
+        let isa = isa_builder.finish(settings::Flags::new(flag_builder));
+        let builder = JITBuilder::with_isa(isa, cranelift_module::default_libcall_names());
         let module = JITModule::new(builder);
         Self {
             builder_context: FunctionBuilderContext::new(),
```
```
λ  (main|✚ 2) cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/toy`
the answer is: 42
recursive_fib(10) = 55
iterative_fib(10) = 55
hello world!
```

I could maybe PR that because I don't see how PIC is really relevant to the demo and it would make it work on more platforms

**Comment by bnjbvr** [comment]
I don't know how PLT tables work in MacOS/aarch64 either. It's easy to make it work for our use case (load from pointer into reg + branch into it), but if the linker expects a precise sequence like the one from @bjorn3's comment, we'd need to find what is precisely expected. I don't plan to work on this immediately, yet I'd be happy to try patches if that can help!

**Comment by bjorn3** [comment]
In case of `cranelift-jit` no linker is involved. Or rather `cranelift-jit` is kind of the linker itself. It is not necessary to copy the exact same sequence. (Even on x86_64 I didn't use the exact same sequence) I merely posted it to show what it would roughly need to look like. Basically all the PLT entry needs to do in the case of `cranelift-jit` on any platform is load the corresponding GOT entry and jump to the loaded address. This is independent of the OS, but once Rust supports pointer authentication on the M1, the PLT entry code will need to be adapted to support pointer authentication.

**Comment by otaviopace** [comment]
Is this related to: https://github.com/bytecodealliance/wasmtime-go/issues/53?

**Comment by cfallin** [comment]
@octaviopace no, this issue doesn't need to be solved in order to run Wasmtime on macOS/aarch64 (M1) -- the `cranelift-jit` crate isn't used by Wasmtime.

**Comment by akirilov-arm** [comment]
Issue #2907 is related to this.

**Comment by Somfic** [comment]
hi, are there any updates on this?  i’m running into

```
PLT is currently only supported on x86_64
```

with `use_colocated_libcalls` and `is_pic` flags in my isa set to `false`

**Comment by cfallin** [comment]
@Somfic no, this issue hasn't been worked on. PRs welcome if you're interested in tackling it!

**Comment by Somfic** [comment]
> @Somfic no, this issue hasn't been worked on. PRs welcome if you're interested in tackling it!

@cfallin I wish I had the technical knowledge to contribute.

As mentioned earlier in this issue thread, setting `is_pic` to `false` should disable PLT entries for aarch64, but it doesn't? Is this a bug or intented behaviour?

**Comment by cfallin** [comment]
cc @bjorn3 -- I don't know the PLT functionality well enough to say, sorry.

**Comment by bjorn3** [comment]
It is used for the hot code swapping functionality of cranelift-jit that I introduced for cg_clif. I've removed hot code swapping support in https://github.com/rust-lang/rustc_codegen_cranelift/commit/5d03df9431833ad992c1507fedc7daee3e232443. I might reintroduce it at some point, but will likely do so by replacing cranelift-jit usage entirely. Because of this and the fact that it reduces portability of cranelift-jit I did propose removing the hot code swapping support from cranelift-jit.

**Comment by bjorn3** [comment]
With https://github.com/bytecodealliance/wasmtime/pull/10345 cranelift-jit should work if you set `is_pic=false` in your `TargetIsa`. In a future PR I plan to enforce `is_pic=false` and remove all the PIC handling code that doesn't work on non-x86_64 targets.
